### DISCUSSION TITLE
Creation of admin and content creator/reviewer roles

### DISCUSSION THREAD

**Discussion by githubpiyush** [Sunbird-Lern/Community#46] [Query] [answered]
How can one create various admin roles in creation portal?
(Admin roles :-  **Sourcing Org Admin** and **Contributor Org Admin**)

How can one create various roles such as **Content Creator** and **Content Reviewer** in consumption portal?

If you have any reference document for this query please share it.

You can go through this discussion for the context :-
https://github.com/Sunbird-coKreat/Community/discussions/33

cc - @kiranharidas187 @vijiurs @aks30 @pradeepkumarcm96

**Comment by reshmi-nair** [comment]
@githubpiyush 

User roles can be created using /user/v2/role/assign API. ORG_ADMIN role has to be assigned internally. All other roles can be assign by user who as ORG_ADMIN role. In request, pass the userId to which the role needs to be assigned and the list of roles. Listt of roles should include the role that needs to be assigned and scope should have the organisation in which the user has that role.
Sample request format:
{
    "request": {
        "userId": "<userid to which role need to be assigned>",
        "roles": [{
				"role": "COURSE_CREATOR", // role that needs to be added
                                "operation":"add",
				"scope": [{
						"organisationId": "<organisation id in which the user will have the above role>"
					}
				]
			}
		]
    }
}
Note: To add roles to a user "operation":"add" need to be mentioned along with role and scope. To remove already assigned roles from a user "operation":"remove" need to be added in the request.

API documentation: [Assign roles](https://lern.sunbird.org/learn/product-and-developer-guide/user-and-org-service/api-documentation/user-management#user-v2-role-assign)

**Reply by githubpiyush** [reply]
@reshmi-nair

I can see various create user API in the documentation, Can you help us on the step by step procedure so that we know which APIs to use.The goal is how can we create a user and assign these roles, in the staging environment?

How to create an organisation, I can not find any API in the org management tab.

We can not create ORG_ADMIN user using the /user/v2/role/assign API, right?

There are two ways to assign content creator/reviewer roles :-
1. User with ORG_ADMIN role can assign using the VDN portal.
2. We can use the API to assign roles to the user.
Is this understanding is correct?

**Reply by reshmi-nair** [reply]
@githubpiyush Pls user Create SsoUser V1 API from  the documentation to create user.
[Create SsoUser V1](https://lern.sunbird.org/learn/product-and-developer-guide/user-and-org-service/api-documentation/user-management#user-v1-sso-create
)
You can pass "roles: ['ORG_ADMIN']" in the above api request to create a user with ORG_ADMIN role.

About assign role what you have mentioned above is correct. Pls refer to [User Roles Documentation](https://lern.sunbird.org/learn/product-and-developer-guide/user-and-org-service/roles/user-roles)

**Reply by githubpiyush** [reply]
@reshmi-nair While assigning roles to users using APIs, we just need the staging authorization token; does that mean anyone (who is not ORG_ADMIN) can assign roles to the user using APIs?

**Reply by reshmi-nair** [reply]
@githubpiyush For assign role API call, x-authenticated-user-token is required, which will have the user roles. This will be checked in the OPA layer. So users who as ORG_ADMIN role only can assign roles

**Reply by githubpiyush** [reply]
> @githubpiyush Pls user Create SsoUser V1 API from the documentation to create user.
> [Create SsoUser V1](https://lern.sunbird.org/learn/product-and-developer-guide/user-and-org-service/api-documentation/user-management#user-v1-sso-create)
> You can pass "roles: ['ORG_ADMIN']" in the above api request to create a user with ORG_ADMIN role.

I am getting timeout in this API, please check the curl shared in the slack.

**Reply by githubpiyush** [reply]
@reshmi-nair Have you checked the timeout error of the create sso user API?

**Reply by Hari-stackroute** [reply]
@githubpiyush "https://dev.open-sunbird.org/" the domain is used is not in use.
where do you get this from?
cc : @reshmi-nair

**Reply by githubpiyush** [reply]
@Hari-stackroute 
We wanted to try user creation in the dev environment, so can you provide me dev API host for the same? 
CC :- @reshmi-nair

**Reply by Hari-stackroute** [reply, marked as answer]
@githubpiyush the following is the curl for non-ed dev environment.
`curl --location --request POST 'https://dev.lern.sunbird.org/api/user/v1/sso/create' \
--header 'Content-Type: application/json' \
--data-raw '{
    "request": {
        "firstName": "custchanneladmin1",
        "password": "Password@1",
        "email": "custchanneladmin1@yopmail.com",
        "userName": "custchanneladmin1",
        "lastName": "custchanneladmin1",
        "emailVerified":true,
        "organisationId":"01269878797503692810",
        "channel":"tn",
        "roles": ["ORG_ADMIN"]
    }
}'`
